server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureService.java

```diff
@@ -0,0 +1,249 @@
+            threadPool,
+            System::nanoTime,
+            List.of(
+                new NodeResourceUsageTracker(
+                    () -> ProcessProbe.getInstance().getProcessCpuPercent() / 100.0 >= settings.getNodeDuressCpuThreshold()
+                ),
+                new NodeResourceUsageTracker(
+                    () -> JvmStats.jvmStats().getMem().getHeapUsedPercent() / 100.0 >= settings.getNodeDuressHeapThreshold()
+                )
+            ),
+            Collections.emptyList()
+        );
+    }
+
+    public SearchBackpressureService(
+        SearchBackpressureSettings settings,
+        TaskResourceTrackingService taskResourceTrackingService,
+        ThreadPool threadPool,
+        LongSupplier timeNanosSupplier,
+        List<NodeResourceUsageTracker> nodeResourceUsageTrackers,
+        List<TaskResourceUsageTracker> taskResourceUsageTrackers
+    ) {
+        this.settings = settings;
+        this.taskResourceTrackingService = taskResourceTrackingService;
+        this.taskResourceTrackingService.addTaskCompletionListener(this);
+        this.timeNanosSupplier = timeNanosSupplier;
+        this.nodeResourceUsageTrackers = nodeResourceUsageTrackers;
+        this.taskResourceUsageTrackers = taskResourceUsageTrackers;
+
+        this.taskCancellationRateLimiter = new TokenBucket(
+            timeNanosSupplier,
+            getSettings().getCancellationRateNanos(),
+            getSettings().getCancellationBurst()
+        );
+
+        this.taskCancellationRatioLimiter = new TokenBucket(
+            state::getCompletionCount,
+            getSettings().getCancellationRatio(),
+            getSettings().getCancellationBurst()
+        );
+
+        threadPool.scheduleWithFixedDelay(this, getSettings().getInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    public void run() {
+        try {
+            doRun();
+        } catch (Exception e) {
+            logger.debug("failure in search search backpressure", e);
+        }
+    }
+
+    public void doRun() {
+        if (getSettings().isEnabled() == false) {
+            return;
+        }
+
+        if (isNodeInDuress() == false) {
+            return;
+        }
+
+        // We are only targeting in-flight cancellation of SearchShardTask for now.
+        List<CancellableTask> searchShardTasks = getSearchShardTasks();
+
+        // Force-refresh usage stats of these tasks before making a cancellation decision.
+        taskResourceTrackingService.refreshResourceStats(searchShardTasks.toArray(new Task[0]));
+
+        // Skip cancellation if the increase in heap usage is not due to search requests.
+        long runningTasksHeapUsage = searchShardTasks.stream().mapToLong(task -> task.getTotalResourceStats().getMemoryInBytes()).sum();
+        if (runningTasksHeapUsage < getSettings().getSearchHeapThresholdBytes()) {
+            return;
+        }
+
+        for (TaskCancellation taskCancellation : getTaskCancellations(searchShardTasks)) {
```

Comment:
I feel it would also be useful to log how much is the heap memory that we are trying to reclaim by cancelling the tasks in every iteration. Is it possible to also log the same once the tasks are chosen for cancellation? Or is there any other means to get that info?